### PR TITLE
core[patch]: Agent Observations should be cast to string to avoid errors

### PR DESCRIPTION
**Pull request by keenborder786**
- **Description:**  When the user calls `astream_events` for `AgentExecutor`, it executes `__aiter__`, and for every chunk it gets from `agent_executor`, it calls its `messages` property. However, when the `messages` property of `AgentStep` is called, the following function is invoked: `_convert_agent_observation_to_messages`, which converts observations to `HumanMessage`. The problem is that `HumanMessage` only expects a string in its content, which was not happening, and therefore the error was raised, as pointed out in the issue below.
    - **Issue:** #26763

### CODE REVIEW

**Comment by keenborder786** [comment]
@eyurtsev

File: libs/core/langchain_core/agents.py [hunk 1]
```diff
@@ -189,10 +189,18 @@ def _convert_agent_observation_to_messages(
...
         AIMessage that corresponds to the original tool invocation.
     """
+
     if isinstance(agent_action, AgentActionMessageLog):
         return [_create_function_message(agent_action, observation)]
     else:
-        return [HumanMessage(content=observation)]
+        if not isinstance(observation, str):
+            try:
+                content = json.dumps(observation, ensure_ascii=False)
```
Comment (eyurtsev) [inline libs/core/langchain_core/agents.py:198]: Could you add a comment to explain why json.dumps is chosen as the default strategy vs. str? ( I'm actually not sure which one should be used -- is there a good way to think about this?)

Comment (keenborder786) [reply libs/core/langchain_core/agents.py:198]: okay let me think about this.

Comment (sharrajesh) [reply libs/core/langchain_core/agents.py:198]: fwiw i added these gists to repro the issue

https://github.com/langchain-ai/langchain/issues/26763#issuecomment-2367315600

this issue happens with tools with dict when agent was created with create_react_agent https://gist.github.com/sharrajesh/765c0b6edfe991363675f45d467e3c93

but does not happen with other variant of agent creation create_openai_tools_agent which i cannot change without introducing risk
https://gist.github.com/sharrajesh/1080af5a95ae9d7b83a8da46597b68e1

i can give you branch a try to make sure it really works if it helps?

Comment (keenborder786) [reply libs/core/langchain_core/agents.py:198]: Yes don't worry @sharrajesh all of these will get resolved.

**Comment by eyurtsev** [comment]
@keenborder786 benchmarking total time isn't relevant (at least not directly). If the encoding affected the number of cycles that the agent had to do because it affected the chat models ability to correctly use the encoded output, then sure that's a relevant consideration, but that's not what's measured in these benchmarks. 

The way to think about json.dumps vs. str is based on:  For things typically returned from tools, how should that output be cast into string format to be best consumed by a chat model. This mean that the chat model would need to act on that information correctly.

Why did you chose json.dumps for this PR? Is there another part in langchain code that has similar logic?

---

As an side: The benchmarks themselves are completely dominated by noise from fluctuations in chat model run times. 

If you care you can verify this by calculating the standard error of the mean (which is the standard deviation divded by the square root of the number of observations) -- https://en.wikipedia.org/wiki/Standard_error, it'll give you a sense about the uncertainty in our measurement of the mean.

**Comment by sharrajesh** [comment]
please also try running create_react_agent version of this code which is causing the issue 

https://gist.github.com/sharrajesh/765c0b6edfe991363675f45d467e3c93

**Comment by keenborder786** [comment]
> @keenborder786 benchmarking total time isn't relevant (at least not directly). If the encoding affected the number of cycles that the agent had to do because it affected the chat models ability to correctly use the encoded output, then sure that's a relevant consideration, but that's not what's measured in these benchmarks.
> 
> The way to think about json.dumps vs. str is based on: For things typically returned from tools, how should that output be cast into string format to be best consumed by a chat model. This mean that the chat model would need to act on that information correctly.
> 
> Why did you chose json.dumps for this PR? Is there another part in langchain code that has similar logic?
> 
> As an side: The benchmarks themselves are completely dominated by noise from fluctuations in chat model run times.
> 
> If you care you can verify this by calculating the standard error of the mean (which is the standard deviation divded by the square root of the number of observations) -- https://en.wikipedia.org/wiki/Standard_error, it'll give you a sense about the uncertainty in our measurement of the mean.

@eyurtsev okay

**Comment by keenborder786** [comment]
> > @keenborder786 benchmarking total time isn't relevant (at least not directly). If the encoding affected the number of cycles that the agent had to do because it affected the chat models ability to correctly use the encoded output, then sure that's a relevant consideration, but that's not what's measured in these benchmarks.
> > The way to think about json.dumps vs. str is based on: For things typically returned from tools, how should that output be cast into string format to be best consumed by a chat model. This mean that the chat model would need to act on that information correctly.
> > Why did you chose json.dumps for this PR? Is there another part in langchain code that has similar logic?
> > As an side: The benchmarks themselves are completely dominated by noise from fluctuations in chat model run times.
> > If you care you can verify this by calculating the standard error of the mean (which is the standard deviation divded by the square root of the number of observations) -- https://en.wikipedia.org/wiki/Standard_error, it'll give you a sense about the uncertainty in our measurement of the mean.
> 

> @eyurtsev . 

-  You are right but then I guess benchmarking is actually futile, because if our concern is `how chat model would need to act on that information correctly` , then we can always use `json.dumps`, and if it throws an error then we can use `str` to cast the obervation.
- Yes I choosed json.dumps because of `_create_function_message` but without the exception but I guess we can use the exact same logic when casting: https://github.com/langchain-ai/langchain/blob/b2c315997c5a3b0664797ae712e23f69f9b5f1ab/libs/core/langchain_core/agents.py#L212, It should being used by `_create_function_message`.

**Comment by eyurtsev** [comment]
> Yes I choosed json.dumps because of _create_function_message but without the exception but I guess we can use the exact same logic when casting (I just made a commit, please see):

Got it -- I didn't see it from the github diff

**Review by eyurtsev — APPROVED** [review]

File: libs/core/langchain_core/agents.py (hunk 1 above)
Comment (sharrajesh) [reply libs/core/langchain_core/agents.py:199]: Team, when you do expect this fix will be merged?

**Pull request merged**